### ISSUE TITLE
Found two critical types of security vulnerabilities

### ISSUE THREAD

**Issue by fl4x** [totaljs/eshop#64]
Recently google ads rejected the website based on eshop due to the presence of malware.
I have tested https://eshop.totaljs.com/ using Burp Suite and it finds two major vulnerabilities:
1. JavaScript injection (DOM-based)
2. Cross-site scripting (DOM-based)

Can someone provide an explanation for a code fix for these vulnerabilities?
See the screenshots attached below.

- List of all issues
![all](https://user-images.githubusercontent.com/10885321/119122138-dfc28680-ba36-11eb-89d5-48acdbbab714.jpg)

- Requests
![1](https://user-images.githubusercontent.com/10885321/119122344-17c9c980-ba37-11eb-9dbf-d030a7cc0b5e.jpg)
![2](https://user-images.githubusercontent.com/10885321/119122364-1b5d5080-ba37-11eb-9156-d41535f1f6d2.jpg)
![3](https://user-images.githubusercontent.com/10885321/119122370-1d271400-ba37-11eb-88de-4b8de53e46b7.jpg)
![4](https://user-images.githubusercontent.com/10885321/119122377-1e584100-ba37-11eb-922f-aa5e8738472e.jpg)
![5](https://user-images.githubusercontent.com/10885321/119122380-20220480-ba37-11eb-9721-67df7cdfb3be.jpg)

**Comment by petersirka** [comment]
Thank you a lot. We will solve it in short time.

**Comment by petersirka** [comment]
I have looked into your results and I can't simulate them correctly. Can you send me here a link with that injection? I can test it.

**Comment by fl4x** [comment]
GET /categories/?q=se4jx08lmpln1oxmr3akjugu1tuzlgkj19q HTTP/1.1
Host: eshop.totaljs.com
Cookie: visit=162186508294245092
Upgrade-Insecure-Requests: 1
Referer: https://eshop.totaljs.com/categories/?q=836324
Accept: */*
Accept-Language: en-US,en-GB;q=0.9,en;q=0.8
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36
Connection: close
Cache-Control: max-age=0
Accept-Encoding: gzip, deflate

![adv1](https://user-images.githubusercontent.com/10885321/119364315-6daca480-bcb7-11eb-9c9d-ba0622c3a806.jpg)

**Comment by fl4x** [comment]
GET /categories/?q=se4jx08lmpln1oxmr3akjugu1tuzlgkj19q HTTP/1.1
Host: eshop.totaljs.com
Cookie: visit=162186508294245092
Upgrade-Insecure-Requests: 1
Referer: https://eshop.totaljs.com/categories/?q=836324
Accept: */*
Accept-Language: en-US,en-GB;q=0.9,en;q=0.8
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36
Connection: close
Cache-Control: max-age=0
Accept-Encoding: gzip, deflate

![adv2](https://user-images.githubusercontent.com/10885321/119364502-a5b3e780-bcb7-11eb-8b79-d8c9b1c28d43.jpg)

**Comment by fl4x** [comment]
GET /categories/?q=pyfhhbsw605ylzhxbeuv3505l4ea5r1tij7&d11ep%22%3e%3ca%3eebr6w=1 HTTP/1.1
Host: eshop.totaljs.com
Cookie: visit=162186533242193706
Upgrade-Insecure-Requests: 1
Referer: https://eshop.totaljs.com/categories/?q=pyfhhbsw605ylzhxbeuv3505l4ea5r1tij7
Accept: */*
Accept-Language: en-US,en-GB;q=0.9,en;q=0.8
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36
Connection: close
Cache-Control: max-age=0
Accept-Encoding: gzip, deflate

**Comment by fl4x** [comment]
GET /robots.txt HTTP/1.1
Host: eshop.totaljs.com
Upgrade-Insecure-Requests: 1
Accept: */*
Accept-Language: en-US,en-GB;q=0.9,en;q=0.8
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36
Connection: close
Cache-Control: max-age=0
Accept-Encoding: gzip, deflate

![adv3_1](https://user-images.githubusercontent.com/10885321/119364817-faeff900-bcb7-11eb-9bc4-602c97756080.jpg)
![adv3_2](https://user-images.githubusercontent.com/10885321/119364820-fcb9bc80-bcb7-11eb-8eec-fbfddd2c318b.jpg)

**Comment by fl4x** [comment]
GET /robots.txt HTTP/1.1
Host: eshop.totaljs.com
Upgrade-Insecure-Requests: 1
Accept: */*
Accept-Language: en-US,en-GB;q=0.9,en;q=0.8
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36
Connection: close
Cache-Control: max-age=0
Accept-Encoding: gzip, deflate

![adv4_1](https://user-images.githubusercontent.com/10885321/119365031-34c0ff80-bcb8-11eb-8e25-cd31530a1c93.jpg)
![adv4_2](https://user-images.githubusercontent.com/10885321/119365047-38548680-bcb8-11eb-8ab6-ec790ed240aa.jpg)

**Comment by fl4x** [comment]
> GET /categories/?q=se4jx08lmpln1oxmr3akjugu1tuzlgkj19q HTTP/1.1
> Host: eshop.totaljs.com
> Cookie: visit=162186508294245092
> Upgrade-Insecure-Requests: 1
> Referer: https://eshop.totaljs.com/categories/?q=836324
> Accept: _/_
> Accept-Language: en-US,en-GB;q=0.9,en;q=0.8
> User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36
> Connection: close
> Cache-Control: max-age=0
> Accept-Encoding: gzip, deflate
> 
> ![adv1](https://user-images.githubusercontent.com/10885321/119364315-6daca480-bcb7-11eb-9c9d-ba0622c3a806.jpg)

This vulnerability appears for each new category created using eshop, so the issue might be in a template code.

**Comment by petersirka** [comment]
I have tested a lot of requests and really I can't simulate that problem. I have tested it via browser, via PAW app and directly via Total.js framework. 

- My requests: [requests.txt](https://github.com/totaljs/eshop/files/6534220/requests.txt)

Can you prepare requests similarly? I want to test it directly, I don't have software like you, commercial version doesn't contain a simple way how to test our website.

Thank you for your time!

**Comment by fl4x** [comment]
I have marked errors that seat in responses with red color. See the MS Word document attached.
For the **JavaScript injection (DOM-based)**, if you need to know exactly which function produces the error, provide me please with the unminified sourcecode for `spa.min@14.js`
[E-Shop vulnerabilities.docx](https://github.com/totaljs/eshop/files/6535019/E-Shop.vulnerabilities.docx)

**Comment by petersirka** [comment]
@fl4x yes, you're right. I have fixed that problem in the Total.js framework because the framework didn't escape the parsed __keys__ from the query argument. You can try my fix on <https://eshop.totaljs.com>.

> provide me please with the unminified sourcecode for spa.min@14.js

The source below is jComponent v18, but the method is very similar. I don't see any problem in `location.pathname` or in `location.href` because the browser encodes all characters. Same conclusion with error page if you call a request to the `robot.txt` files. Can you explain to me how I can simulate that mentioned security bug?

- https://github.com/totaljs/jComponent/blob/master/jc.js#L1819

### Thank you a lot!

**Comment by fl4x** [comment]
Hi @petersirka. Could you please add the changes you've applied to https://eshop.totaljs.com/ to the current master branch, or point what have you changed, because the ` 3. Cross-site scripting (reflected)` vulnerability has been fixed.

Concerning ` spa.min@14.js - location.pathname` and  `https://eshop.totaljs.com/robots.txt - location.href` I am not a hacker trying to hack your website, thus I can not spend much time, trying to construct a harmful request. I am just trying to find a way to fix all the vulnerabilities with high priority, so the eshop would be free of them and not detected by BurpSuite.

Is there a way to avoid using `location.pathname` and `location.href`?

**Comment by petersirka** [comment]
@fl4x thank you so much for your time. I appreciate it.

> Could you please add the changes you've applied to

We work on a new version of Eshop, but the fix is part of the Total.js framework (not the part of Eshop). I have added the link to your GitHub profile to our monthly blog post reports:

![image](https://user-images.githubusercontent.com/2414252/119479775-47d3de00-bd51-11eb-8c5e-fb8fa3761300.png)

If you want, I can change a link to your website.

> Is there a way to avoid using location.pathname and location.href?

It's a bit problem because `location.pathname` is using in the jComponent library; I can't remove it. The `location.href` is part of the error page (I can modify/improve this), but I have tested it, and the value is correctly escaped.

---

I'll publish a new release of the framework during this week.

Again: I appreciate your work!
Thank you!
